client/client-game-state-queue.cpp

```diff
@@ -0,0 +1,233 @@
+				if (base_orientation_iter != this->base_state.orientations.end()) {
+					this->interpolated_state.orientations[orientation.first].value =
+						glm::slerp(base_orientation_iter->second.value, orientation.second.value, lerp_percent);
+				}
+				else {
+					this->interpolated_state.orientations[orientation.first] = orientation.second;
+				}
+			}
+		}
+	}
+}
+
+void ClientGameStateQueue::ProcessEventQueue() {
+	EventQueue<EntityCreated>::ProcessEventQueue();
+	EventQueue<EntityDestroyed>::ProcessEventQueue();
+	EventQueue<NewGameStateEvent>::ProcessEventQueue();
+}
+
+void ClientGameStateQueue::On(std::shared_ptr<NewGameStateEvent> data) {
+	QueueServerState(std::move(data->new_state));
+}
+
+void ClientGameStateQueue::QueueServerState(GameState&& new_state) {
+	if (new_state.state_id > this->last_server_state_id) {
+		this->server_states_array[server_state_array_index % SERVER_STATES_ARRAY_SIZE] = new_state;
+		server_state_array_index++;
+		this->last_server_state_id = new_state.state_id;
+		std::lock_guard<std::mutex> lg(this->server_state_mutex);
+		CheckPredictionResult(new_state);
+		this->server_states.emplace(std::move(new_state));
+	}
+}
+
+void ClientGameStateQueue::CheckPredictionResult(GameState& new_state) {
+	if (this->client_id == 0) {
+		return;
+	}
+	this->stats.current_acked_id = new_state.command_id;
+	// Buffer past one incoming state
+	// FIXME we don't actually know when the server applies the commands it ACKs!
+	// so which state we should consider for the error reference is highly ping dependant
+	// this would require changes to how the server reports movement commands
+	// for now, waiting one state seems good enough, at least on localhost, over the net is different...
+	glm::vec3 state_pos = this->stats.server_position_next;
+	this->stats.server_position = this->stats.server_position_next;
+	this->stats.server_position_next = new_state.positions[this->client_id].value;
+
+	glm::vec3 position_diff = glm::vec3();
+	glm::vec3 velocity_diff = glm::vec3();
+
+	// delete from the prediction queue anything that's not newer than the latest ack
+	for (auto itr = this->predictions.begin(); itr != this->predictions.end(); ) {
+		if (itr->first <= new_state.command_id) {
+			// if this was an acked command, calculate the difference between where we thought we were
+			// this gives feedback on how good our prediction was
+			if (itr->first == new_state.command_id) {
+				glm::vec3 predict_pos = itr->second.positions[this->client_id].value;
+				position_diff = state_pos - predict_pos;
+				glm::vec3 state_vel = new_state.velocities[this->client_id].linear;
+				glm::vec3 predict_vel = itr->second.velocities[this->client_id].linear;
+				velocity_diff = state_vel - predict_vel;
+			}
+			itr = this->predictions.erase(itr);
+			continue;
+		}
+		++itr;
+	}
+	// debugging stats, also used by reconciliation
+	stats.client_position = position_diff;
```

Comment:
The client_position variable name is a bit misleading

Reply:
After having refactored this a few times, I would agree.